test/integration/http_integration.cc

```diff
@@ -973,6 +973,78 @@ void HttpIntegrationTest::testRouterUpstreamResponseBeforeRequestComplete() {
   EXPECT_EQ(512U, response->body().size());
 }
 
+void HttpIntegrationTest::testRouterRetryOnResetBeforeRequestAfterHeaders() {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto encoder_decoder = codec_client_->startRequest(
+      Http::TestRequestHeaderMapImpl{{":method", "POST"},
+                                     {":path", "/test/long/url"},
+                                     {":scheme", "http"},
+                                     {":authority", "sni.lyft.com"},
+                                     {"x-forwarded-for", "10.0.0.1"},
+                                     {"x-envoy-retry-on", "reset-before-request"}});
+  waitForNextUpstreamConnection(std::vector<uint64_t>({0}), TestUtility::DefaultTimeout,
+                                fake_upstream_connection_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+  codec_client_->sendData(*request_encoder_, 1024, true);
+  waitForNextUpstreamRequest();
+  // Reset the upstream connection after the headers have been sent
+  ASSERT_TRUE(fake_upstream_connection_->close());
+
+  // We should get no new stream/retry
+  ASSERT_FALSE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  cleanupUpstreamAndDownstream();
+}
+
+void HttpIntegrationTest::testRouterRetryOnResetBeforeRequestBeforeHeaders() {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto encoder_decoder = codec_client_->startRequest(
+      Http::TestRequestHeaderMapImpl{{":method", "POST"},
+                                     {":path", "/test/long/url"},
+                                     {":scheme", "http"},
+                                     {":authority", "sni.lyft.com"},
+                                     {"x-forwarded-for", "10.0.0.1"},
+                                     {"x-envoy-retry-on", "reset-before-request"}});
+  waitForNextUpstreamConnection(std::vector<uint64_t>({0}), TestUtility::DefaultTimeout,
+                                fake_upstream_connection_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+  // Reset the upstream connection before the headers have been sent
```

Comment:
ah so not waiting for the next upstream connection doesn't mean the headers aren't sent.  I think you need a pause filter to ensure the headers haven't been shipped upstream.  If you startRequest and use the buffer filter that should do it.

If this is passing consistently I'm actually really confused how.
hm, maybe it's because you're not setting upstream_request_started_(false) in the filter constructor?  Can you check and see if doing that causes this test to fail?

Reply:
Can I do that by prepending the "buffer-continue-filter" as an upstream filter?

Reply:
anything that pauses encoding headers should work